blockscout-ens/bens-logic/src/subgraph/ens/wildcard.rs

```diff
@@ -0,0 +1,186 @@
+) -> cached::Return<Option<CreationDomain>> {
+    metrics::WILDCARD_RESOLVE_ATTEMPTS.inc();
+    match try_wildcard_resolution(db, from_user).await {
+        Ok(result) => {
+            if result.is_some() {
+                metrics::WILDCARD_RESOLVE_SUCCESS.inc();
+            }
+            cached::Return::new(result)
+        }
+        Err(err) => {
+            tracing::error!(
+                name = from_user.inner.name,
+                error = %err,
+                "error while trying wildcard resolution"
+            );
+            cached::Return::new(None)
+        }
+    }
+}
+
+async fn try_wildcard_resolution(
+    db: &PgPool,
+    from_user: &DomainNameOnProtocol<'_>,
+) -> Result<Option<CreationDomain>, anyhow::Error> {
+    let Some((resolver_address, maybe_existing_domain)) = any_resolver(db, from_user).await? else {
+        return Ok(None);
+    };
+
+    if let Some(domain) = &maybe_existing_domain {
+        if domain.id == from_user.inner.id {
+            if !domain.resolved_with_wildcard {
+                // we found domain that resolved by graph node already, skip it
+                return Ok(None);
+            } else {
+                // domain is resolved with wildcard, so we need to recheck it
+                tracing::info!(
+                    domain_id = domain.id,
+                    domain_name = domain.name,
+                    "found domain with wildcard-resolution, rechecking it"
+                )
+            }
+        }
+    };
+
+    let result = ccip_read::call_to_resolver(from_user, resolver_address)
+        .await
+        .context("resolve using ccip call")?;
+
+    if !result.addr.is_zero() {
+        Ok(Some(creation_domain_from_rpc_resolution(
+            from_user,
+            result,
+            maybe_existing_domain,
+        )))
+    } else {
+        Ok(None)
+    }
+}
+
+async fn any_resolver(
+    db: &PgPool,
+    from_user: &DomainNameOnProtocol<'_>,
+) -> Result<Option<(Address, Option<Domain>)>, anyhow::Error> {
+    let protocol = from_user.deployed_protocol.protocol;
+    let name_options = from_user
+        .inner
+        .iter_parents_with_self()
+        .map(|name| DomainNameOnProtocol::new(name, from_user.deployed_protocol))
+        .collect::<Vec<_>>();
+    // try to find resolver in db
+    let maybe_domain_with_resolver = any_resolver_in_db(db, name_options.clone()).await?;
+    if let Some((resolver, found_domain)) = maybe_domain_with_resolver {
+        Ok(Some((resolver.resolver_address, Some(found_domain))))
+    } else if protocol.info.registry_contract.is_some() || protocol.info.network_id == 1 {
+        // try to find resolver on chain.
+        // if custom registry is set, we can try to find resolver in registry
+        if let Some(resolver_address) = any_resolver_rpc(name_options).await {
+            Ok(Some((resolver_address, None)))
+        } else {
+            Ok(None)
+        }
+    } else {
+        Ok(None)
+    }
```

Comment:
Nit: maybe avoid nesting of if else statements

```suggestion
    if let Some((resolver, found_domain)) = maybe_domain_with_resolver {
        return Ok(Some((resolver.resolver_address, Some(found_domain))))
    } else if protocol.info.registry_contract.is_some() || protocol.info.network_id == 1 {
        // try to find resolver on chain.
        // if custom registry is set, we can try to find resolver in registry
        if let Some(resolver_address) = any_resolver_rpc(name_options).await {
            return Ok(Some((resolver_address, None)))
        }
    }
    Ok(None)
```